collector/nvidia.go

```diff
@@ -0,0 +1,305 @@
+	if C.nvmlDeviceGetHandleByIndex_v2Ptr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetNamePtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetName"))
+	if C.nvmlDeviceGetNamePtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetGraphicsRunningProcesses_v2Ptr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetGraphicsRunningProcesses_v2"))
+	if C.nvmlDeviceGetGraphicsRunningProcesses_v2Ptr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlSystemGetProcessNamePtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlSystemGetProcessName"))
+	if C.nvmlSystemGetProcessNamePtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetDriverModelPtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetDriverModel"))
+	if C.nvmlDeviceGetDriverModelPtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	C.nvmlDeviceGetMemoryInfoPtr = C.GetProcAddress(nvmlLibrary, C.CString("nvmlDeviceGetMemoryInfo"))
+	if C.nvmlDeviceGetMemoryInfoPtr == C.FARPROC(C.NULL) {
+		return nil, syscall.Errno(C.GetLastError())
+	}
+
+	err := C.invoke_nvmlInit_v2()
+	if err != C.NVML_SUCCESS {
+		return nil, err
+	}
+
+	const subsystem = "nvidia"
+
+	if *processWhitelist == ".*" && *processBlacklist == "" {
+		log.Warn("No filters specified for nvidia collector. This will generate a very large number of metrics!")
+	}
+
+	var devices []nvdiaGpuDevice
+	if len(*nvidiaGpuWhitelist) > 0 {
+		var count C.uint = 0
+		err := C.invoke_nvmlDeviceGetCount_v2(&count)
+		if err != C.NVML_SUCCESS {
+			return nil, err
+		}
+		var nvidiaGpuWhitelistPatttern = regexp.MustCompile(fmt.Sprintf("^(?:%s)$", *nvidiaGpuWhitelist))
+		for i := C.uint(0); i < count; i++ {
+			var device C.nvmlDevice_t
+			err := C.invoke_nvmlDeviceGetHandleByIndex_v2(i, &device)
+			if err != C.NVML_SUCCESS {
+				continue
+			}
+			nameBuffer := make([]byte, C.NVML_DEVICE_NAME_BUFFER_SIZE)
+			err = C.invoke_nvmlDeviceGetName(device, (*C.char)(unsafe.Pointer(&nameBuffer[0])), C.NVML_DEVICE_NAME_BUFFER_SIZE)
+			if err != C.NVML_SUCCESS {
+				continue
+			}
+			name := string(nameBuffer[:clen(nameBuffer)])
+			var currentDriverModel, pendingDriverModel C.nvmlDriverModel_t
+			err = C.invoke_nvmlDeviceGetDriverModel(device, &currentDriverModel, &pendingDriverModel)
+			if err != C.NVML_SUCCESS {
+				continue
+			}
+			if currentDriverModel == C.NVML_DRIVER_WDDM {
+				log.Warnf("Gpu %s is using WWDM driver that does not allow collecting per process memory usage\n", name)
+			}
+			if nvidiaGpuWhitelistPatttern.MatchString(name) {
+				devices = append(devices, nvdiaGpuDevice{device: device, driverModel: currentDriverModel, name: name})
+			}
+		}
+	}
+
+	return &nvidiaCollector{
+		TotalProcessGpuMemoryUsed: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "total_process_gpu_memory_used"),
```

Comment:
It is good to indicate the unit of measure in the name, so eg `total_process_gpu_memory_used_bytes` (I guess)

Reply:
Ok noted, will fix.